### DISCUSSION TITLE
PostGIS query: Using PostGIS "ST_DWithin" with Javascript API

### DISCUSSION THREAD

**Discussion by bfgpab** [supabase/supabase#20634] [Questions] [answered]
Dear community,

Is it possible to query PostGIS "ST_DWithin" via Javascript API? 

I have column with GPS coordinates called "adr_gps". I want to return rows which are within certain distance from the location specified by user.

Adr_lon, Adr_lat are logitude and latitude of location specified by user in search. Range is the range in meters from the location specified by user in search.

SQL syntax would be: 
` SELECT * FROM 'TABLE NAME' WHERE ST_DWithin(adr_gps::geography, ST_MakePoint(adr_lon, adr_lat)::geography, range))`

Thank you!

**Comment by GaryAustin1** [comment]
You are probably better off with an rpc call to a setof record or table returning function.
Here is the bit of information on the REST API and PostGIS.  https://postgrest.org/en/stable/how-tos/working-with-postgresql-data-types.html#postgis

**Reply by bfgpab** [reply]
Hi Gary, thanks for your response! My PostGIS query works well in RPC, but I actually had difficulties figuring out the way to return values from referenced table. That's why I was looking for a way to query postGIS in Javascript as reference table query in javascript is easy. 

Perhaps you could help me? I have 2 tables, 1 called "ads" and 2nd called "ads_features". I want to return ads with features specified by user in Search filter UI. If in javascript I would call:

`const { data, error } = await supabase.from('ads').select("*, ads_features(feature)").eq("status, "active").in("ads_features.feature", ["feature 1", "feature 2]")` 

Would I replicate same in RPC with:

```
SELECT *, array_agg(ads_features.feature) as ads_features
FROM ads
INNER JOIN ads_features ON ads_features.ad = ads.id
WHERE
    (ads.status = "active") AND
    (ads_features.feature = ANY(ARRAY["feature 1", "feature 2"]))
GROUP BY ads.id;
```
? 
Thanks!

If above query is correct, then I would club it with my location query (posted separately for community reference) and RPC would do the job for me :)

**Reply by bfgpab** [reply, marked as answer]
For someone who might be interested, here is the RPC function that works:

```
CREATE OR REPLACE FUNCTION FilterAdsByLocation(
    adr_lat numeric default NULL,
    adr_lon numeric default NULL,
    range numeric default NULL
)
RETURNS setof ads
LANGUAGE plpgsql
AS $$
BEGIN
    RETURN QUERY
    SELECT *
    FROM ads
    WHERE (adr_lat IS NULL OR adr_lon IS NULL OR ST_DWithin(ads.adr_gps::geography, ST_MakePoint(adr_lon, adr_lat)::geography, range));
END;
$$;
```
